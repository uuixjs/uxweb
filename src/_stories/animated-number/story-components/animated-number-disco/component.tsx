import * as React from "react";

import { AnimatedNumber, Layout } from "v2";

import { Component } from "react";

interface DiscoProps {
  startValue: number;
  deltaMin?: number;
  deltaMax?: number;
}

interface DiscoState {
  value: number;
}

export class AnimatedNumberDisco extends Component<DiscoProps, DiscoState> {
  private interval: number | undefined;

  public constructor(props: DiscoProps) {
    super(props);
    this.state = {
      value: props.startValue,
    };
  }

  public componentDidMount() {
    this.interval = window.setInterval(this.tick, 2000);
  }

  public componentWillUnmount() {
    window.clearInterval(this.interval);
  }

  public render() {
    return (
      <Layout>
        <AnimatedNumber
          value={this.state.value}
          formatValue={(n) => n.toLocaleString()}
        />
      </Layout>
    );
  }

  private tick = () => {
    const diff = generateRandomInteger(
      this.props.deltaMin || -10,
      this.props.deltaMax || 10,
    );

    this.setState((prevState) => ({
      value: prevState.value + diff,
    }));
  };
}

function generateRandomInteger(min: number, max: number) {
  return Math.floor(min + Math.random() * (max + 1 - min));
}
