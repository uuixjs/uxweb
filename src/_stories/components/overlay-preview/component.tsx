import "./styles.scss";

import * as React from "react";

import {
  AlignItems,
  Background,
  CoreImage,
  Display,
  InjectLayout,
  Layout,
  Overflow,
  OverlayRegion,
  Position,
  Radio,
  TransitionGroup,
  TransitionType,
  ZIndex,
} from "v2";
import { Component, FormEvent } from "react";
import { Padding, cn, newUUIDv4 } from "@uuixjs/uuixweb-lib";

import { ExampleThumbnails } from "../../assets";

export enum ExampleBG {
  Off = "non-overlay",
  Photo1 = "photo-1",
  Photo2 = "photo-2",
  Video = "video",
  Black = "black",
  White = "white",
}

interface Props {
  defaultBg?: ExampleBG;
  blur?: boolean;
  padding?: Padding;
}

interface State {
  bg: ExampleBG;
}

export class OverlayPreview extends Component<Props, State> {
  public static defaultProps = {
    padding: 2,
    blur: true,
  };

  private uid = newUUIDv4();

  public constructor(props: Props) {
    super(props);
    this.state = {
      bg: props.defaultBg || ExampleBG.Black,
    };
  }

  public render() {
    const overlay = this.state.bg !== ExampleBG.Off;

    return (
      <Layout>
        <Layout display={Display.Flex} padding={1}>
          {Object.keys(ExampleBG).map((k, i) => (
            <Layout key={k} padding={{ left: i > 0 ? 1 : 0 }}>
              <Radio
                name={`overlay-preview-${this.uid}`}
                label={k}
                value={ExampleBG[k as keyof typeof ExampleBG]}
                onChange={this.onRadioSelected}
                checked={
                  this.state.bg === ExampleBG[k as keyof typeof ExampleBG]
                }
              />
            </Layout>
          ))}
        </Layout>

        <OverlayRegion overlay={overlay}>
          <Layout
            className="overlay-preview__container"
            position={Position.Relative}
            background={overlay ? Background.Overlay : undefined}
            padding={this.props.padding}
          >
            {this.renderBackground()}
            <Layout zIndex={ZIndex.Above} position={Position.Relative}>
              {this.props.children}
            </Layout>
          </Layout>
        </OverlayRegion>
      </Layout>
    );
  }

  private renderBackground() {
    let element;
    if (this.state.bg === ExampleBG.Photo1) {
      element = <CoreImage src={ExampleThumbnails.stream1} alt="" />;
    }

    if (this.state.bg === ExampleBG.Photo2) {
      element = <CoreImage src={ExampleThumbnails.stream2} alt="" />;
    }

    if (this.state.bg === ExampleBG.Video) {
      element = (
        <video
          src="https://d3hmvhl7ru3t12.cloudfront.net/video/pages/home/header-82b9abce067992beedde6d103b6241d616a7ee97cc0d4f3126277098c2e4b09004325f25c6a245656096bd333c880b0032e2cbdbed9d2514220e20718482e95b.mp4"
          autoPlay
          muted
          loop
        />
      );
    }

    if (this.state.bg === ExampleBG.Black) {
      element = <div style={{ background: "#000" }} />;
    }

    if (this.state.bg === ExampleBG.White) {
      element = <div style={{ background: "#FFF" }} />;
    }

    return (
      <InjectLayout
        display={Display.Flex}
        alignItems={AlignItems.Stretch}
        attachTop
        attachLeft
        attachBottom
        attachRight
        overflow={Overflow.Hidden}
      >
        <TransitionGroup
          className="overlay-preview__transition-group-container"
          transitionType={TransitionType.Fade}
        >
          {element ? (
            <InjectLayout
              className={cn({
                ["overlay-preview__bg-element"]: true,
                ["overlay-preview__bg-element--blur"]: this.props.blur,
              })}
              key={this.state.bg}
              position={Position.Absolute}
              attachTop
              attachLeft
              attachBottom
              attachRight
              fullWidth
              fullHeight
            >
              {element}
            </InjectLayout>
          ) : (
            []
          )}
        </TransitionGroup>
      </InjectLayout>
    );
  }

  private onRadioSelected = (e: FormEvent<HTMLInputElement>) => {
    if (e.currentTarget.checked) {
      this.setState({
        bg: e.currentTarget.value as ExampleBG,
      });
    }
  };
}
