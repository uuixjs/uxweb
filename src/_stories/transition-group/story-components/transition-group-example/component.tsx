import * as React from "react";

import {
  AlignItems,
  Background,
  Button,
  ButtonType,
  Column,
  Display,
  FlexDirection,
  FlexWrap,
  FormGroup,
  Grid,
  Input,
  InputType,
  JustifyContent,
  Layout,
  Range,
  Select,
  Tower,
  TowerChildWidth,
  TowerGutter,
  TransitionDuration,
  TransitionGroup,
  TransitionType,
  TransitionTypeOption,
} from "v2";
import { Component, FormEvent, MouseEvent } from "react";

import { TransitionGroupChecklist } from "../transition-group-checklist";
import { newUUIDv4 } from "@uuixjs/uuixweb-lib";

enum DemoLayoutType {
  Ordinary = "ordinary",
  Flexbox = "flexbox",
  // eslint-disable-next-line @typescript-eslint/no-shadow
  Tower = "tower",
}

interface Item {
  key: string;
}

export interface UpdateSettings {
  numToAdd: number;
  addAtIndex: number;
  numToRemove: number;
  removeAtIndex: number;
  numToMove: number;
  moveFromIndex: number;
  moveToIndex: number;
}

export interface LayoutSettings {
  layout: DemoLayoutType;
  direction: FlexDirection;
  duration: TransitionDuration | number;
  transitionType: TransitionTypeOption;
  boxSize: number;
}

interface Props {}

interface State extends UpdateSettings, LayoutSettings {
  showItems: boolean;
  items: Item[];
}

const INITIAL_VALUES: UpdateSettings = {
  numToAdd: 1,
  addAtIndex: 1,
  numToRemove: 0,
  removeAtIndex: 1,
  numToMove: 0,
  moveFromIndex: 0,
  moveToIndex: 1,
};

export class TransitionGroupExample extends Component<Props, State> {
  public state: State = {
    ...INITIAL_VALUES,
    layout: DemoLayoutType.Ordinary,
    direction: FlexDirection.Column,
    duration: TransitionDuration.Long,
    transitionType: [TransitionType.ScaleOver, TransitionType.ScaleOver],
    boxSize: 240,
    showItems: true,
    items: [],
  };

  public componentDidMount() {
    this.init();
  }

  public render() {
    return (
      <Layout margin={{ top: 2 }} border background={Background.Base}>
        <Grid>
          <Column cols={{ default: 12, md: 6, lg: 5, xxl: 4 }}>
            {this.renderListControls()}
          </Column>

          <Column cols={{ default: 12, md: 6, lg: 7, xxl: 8 }}>
            {this.renderLayoutControls()}
          </Column>

          <Column cols={{ default: 12, lg: 12, xxl: 8 }}>
            {this.state.showItems && (
              <div
                style={{
                  display: "block",
                  border: "2px dashed lightgray",
                  margin: "2rem",
                }}
              >
                {this.renderExample()}
              </div>
            )}
          </Column>

          <Column cols={{ default: 12, lg: 12, xxl: 4 }}>
            <Layout padding={2}>
              <TransitionGroupChecklist
                setValues={this.setValues}
                updateView={this.changeItems}
                setItemCount={this.resetItemsWithoutTransition}
                addItem={this.addItem}
                removeItem={this.removeItem}
              />
            </Layout>
          </Column>
        </Grid>
      </Layout>
    );
  }

  private renderExample() {
    if (this.state.layout === DemoLayoutType.Ordinary) {
      return (
        <TransitionGroup
          duration={this.state.duration}
          transitionType={this.state.transitionType}
        >
          {this.renderItems()}
        </TransitionGroup>
      );
    }
    if (this.state.layout === DemoLayoutType.Flexbox) {
      return (
        <TransitionGroup
          duration={this.state.duration}
          transitionType={this.state.transitionType}
          component={
            <Layout
              display={Display.Flex}
              flexDirection={this.state.direction}
              flexWrap={FlexWrap.Wrap}
              padding={1}
            />
          }
        >
          {this.renderItems()}
        </TransitionGroup>
      );
    }

    if (this.state.layout === DemoLayoutType.Tower) {
      return (
        <TransitionGroup
          duration={this.state.duration}
          transitionType={this.state.transitionType}
          component={
            <Tower
              childWidth={TowerChildWidth.Medium}
              gutterSize={TowerGutter.ExtraSmall}
              placeholderItems={20}
            />
          }
        >
          {this.renderItems()}
        </TransitionGroup>
      );
    }
  }

  private renderItems() {
    return this.state.items.map((item, index) => (
      <div
        key={item.key}
        style={{
          display: "flex",
          flexDirection: "column",
          justifyContent: "center",
          alignItems: "center",
          width: `100%`,
          height: this.state.boxSize ? `${this.state.boxSize / 2}px` : "100px",
          background: "green",
          color: "rgba(255,255,255,0.9)",
          fontSize: "1.5rem",
          border: "2px solid darkgreen",
          margin: "",
        }}
      >
        <Layout
          display={Display.Flex}
          flexDirection={FlexDirection.Column}
          alignItems={AlignItems.Center}
          margin={{ bottom: 1 }}
        >
          <div>index: {index}</div>
          <Layout display={Display.Flex} alignItems={AlignItems.Center}>
            <div>key: {item.key.substring(item.key.length - 6)}</div>
            <div
              style={{
                width: "20px",
                height: "20px",
                background: `#${item.key.substring(item.key.length - 6)}`,
                margin: "0 0.5rem",
              }}
            />
          </Layout>
        </Layout>
        <Button
          variant={ButtonType.Secondary}
          onClick={() => this.removeItem(item.key)}
          overlay
        >
          Remove
        </Button>
      </div>
    ));
  }

  private renderListControls = () => {
    return (
      <form onSubmit={this.onSubmitForm}>
        <Layout
          display={Display.InlineFlex}
          flexDirection={FlexDirection.Column}
          alignItems={AlignItems.End}
          padding={2}
        >
          <Layout margin={{ bottom: 1 }}>
            Move
            <div
              style={{
                display: "inline-block",
                width: "50px",
                margin: "0 0.5rem",
              }}
            >
              <Input
                type={InputType.Number}
                value={this.state.numToMove.toString()}
                onChange={this.setNumToMove}
                min={0}
                max={10}
              />
            </div>
            items from
            <div
              style={{
                display: "inline-block",
                width: "50px",
                margin: "0 0.5rem",
              }}
            >
              <Input
                type={InputType.Number}
                value={this.state.moveFromIndex.toString()}
                onChange={this.setMoveFromIndex}
                min={0}
              />
            </div>
            to
            <div
              style={{
                display: "inline-block",
                width: "50px",
                margin: "0 0 0 0.5rem",
              }}
            >
              <Input
                type={InputType.Number}
                value={this.state.moveToIndex.toString()}
                onChange={this.setMoveToIndex}
                min={0}
              />
            </div>
          </Layout>

          <Layout margin={{ top: 0.5 }}>
            Add
            <div
              style={{
                display: "inline-block",
                width: "50px",
                margin: "0 0.5rem",
              }}
            >
              <Input
                type={InputType.Number}
                value={this.state.numToAdd.toString()}
                onChange={this.setNumToAdd}
                min={0}
                max={10}
              />
            </div>
            items at index
            <div
              style={{
                display: "inline-block",
                width: "50px",
                margin: "0 0 0 0.5rem",
              }}
            >
              <Input
                type={InputType.Number}
                value={this.state.addAtIndex.toString()}
                onChange={this.setAddAtIndex}
                min={0}
              />
            </div>
          </Layout>

          <Layout margin={{ top: 0.5 }}>
            Remove
            <div
              style={{
                display: "inline-block",
                width: "50px",
                margin: "0 0.5rem",
              }}
            >
              <Input
                type={InputType.Number}
                value={this.state.numToRemove.toString()}
                onChange={this.setNumToRemove}
                min={0}
                max={10}
              />
            </div>
            items at index
            <div
              style={{
                display: "inline-block",
                width: "50px",
                margin: "0 0 0 0.5rem",
              }}
            >
              <Input
                type={InputType.Number}
                value={this.state.removeAtIndex.toString()}
                onChange={this.setRemoveAtIndex}
                min={0}
              />
            </div>
          </Layout>

          <Layout display={Display.Flex} margin={{ top: 1 }}>
            <Layout margin={{ right: 1 }}>
              <Button onClick={this.onSubmitForm}>Update</Button>
            </Layout>
            <Button onClick={this.init} variant={ButtonType.Secondary}>
              Reset
            </Button>
          </Layout>
        </Layout>
      </form>
    );
  };

  private generateItem = (): Item => {
    return {
      key: newUUIDv4(),
    };
  };

  private init = (e?: MouseEvent<HTMLElement>) => {
    if (e) {
      e.preventDefault();
    }

    this.setState({
      ...INITIAL_VALUES,
      // Initial items have static keys and color values to ensure screenshot diff passes
      items: [{ key: "aa0000" }, { key: "bbbb00" }, { key: "0033cc" }],
    });
  };

  private resetItemsWithoutTransition = async (num: number) => {
    return new Promise<void>((resolve) => {
      this.setState(
        {
          showItems: false,
        },
        () => {
          // wait for full re-render
          requestAnimationFrame(() => {
            this.setState(
              {
                showItems: true,
                items: Array.from(Array(num), this.generateItem),
              },
              resolve,
            );
          });
        },
      );
    });
  };

  private onSubmitForm = (
    e: FormEvent<HTMLFormElement> | MouseEvent<HTMLElement>,
  ) => {
    e.preventDefault();
    this.changeItems();
  };

  private changeItems = () => {
    this.setState((prevState) => {
      const {
        numToAdd,
        addAtIndex,
        numToRemove,
        removeAtIndex,
        numToMove,
        moveFromIndex,
        moveToIndex,
      } = prevState;

      const items = prevState.items.concat();
      const newItems = Array.from(Array(numToAdd), this.generateItem);

      // Remove some
      items.splice(removeAtIndex, numToRemove);

      // Add some
      items.splice(addAtIndex, 0, ...newItems);

      // Move some
      const moving = items.splice(moveFromIndex, numToMove);
      items.splice(moveToIndex, 0, ...moving);

      return {
        items,
      };
    });
  };

  private addItem = (key: string, index: number = 0) => {
    this.setState((prevState) => {
      const items = prevState.items.concat();
      items.splice(index, 0, { key });
      return {
        items,
      };
    });
  };

  private removeItem = (key: string) => {
    this.setState((prevState) => ({
      items: prevState.items.filter((item) => item.key !== key),
    }));
  };

  private setNumToAdd = (e: FormEvent<HTMLInputElement>) => {
    this.setState({
      numToAdd: parseInt(e.currentTarget.value, undefined),
    });
  };

  private setAddAtIndex = (e: FormEvent<HTMLInputElement>) => {
    this.setState({
      addAtIndex: parseInt(e.currentTarget.value, undefined),
    });
  };

  private setNumToRemove = (e: FormEvent<HTMLInputElement>) => {
    this.setState({
      numToRemove: parseInt(e.currentTarget.value, undefined),
    });
  };

  private setRemoveAtIndex = (e: FormEvent<HTMLInputElement>) => {
    this.setState({
      removeAtIndex: parseInt(e.currentTarget.value, undefined),
    });
  };

  private setNumToMove = (e: FormEvent<HTMLInputElement>) => {
    this.setState({
      numToMove: parseInt(e.currentTarget.value, undefined),
    });
  };

  private setMoveFromIndex = (e: FormEvent<HTMLInputElement>) => {
    this.setState({
      moveFromIndex: parseInt(e.currentTarget.value, undefined),
    });
  };

  private setMoveToIndex = (e: FormEvent<HTMLInputElement>) => {
    this.setState({
      moveToIndex: parseInt(e.currentTarget.value, undefined),
    });
  };

  private setValues = async (settings?: Partial<UpdateSettings>) => {
    return new Promise<void>((resolve) =>
      this.setState(
        {
          ...INITIAL_VALUES,
          ...settings,
        },
        resolve,
      ),
    );
  };

  private renderLayoutControls() {
    return (
      <Layout
        padding={2}
        fullWidth
        display={Display.Flex}
        justifyContent={JustifyContent.Start}
        flexWrap={FlexWrap.Wrap}
      >
        <Layout flexGrow={0} margin={{ right: 2 }}>
          <FormGroup label="Layout Demo">
            <Select value={this.state.layout} onChange={this.updateLayout}>
              {Object.keys(DemoLayoutType).map((key: string) => (
                <option
                  value={DemoLayoutType[key as keyof typeof DemoLayoutType]}
                  key={DemoLayoutType[key as keyof typeof DemoLayoutType]}
                >
                  {key}
                </option>
              ))}
            </Select>
          </FormGroup>
        </Layout>

        <Layout flexGrow={0} margin={{ right: 2 }}>
          <FormGroup label="FlexDirection">
            <Select
              value={this.state.direction}
              onChange={this.updateDirection}
              disabled
            >
              <option value={FlexDirection.Row}>Row</option>
              <option value={FlexDirection.Column}>Column</option>
            </Select>
          </FormGroup>
        </Layout>

        <Layout flexGrow={0} margin={{ right: 2 }}>
          <FormGroup label="TransitionDuration">
            <Select
              value={String(this.state.duration)}
              onChange={this.updateDuration}
            >
              {Object.keys(TransitionDuration).map((key: string) => (
                <option
                  value={
                    TransitionDuration[key as keyof typeof TransitionDuration]
                  }
                  key={
                    TransitionDuration[key as keyof typeof TransitionDuration]
                  }
                >
                  {key}
                </option>
              ))}
              <option value="1000" key="1000ms">
                1,000ms
              </option>
              <option value="3000" key="3000ms">
                3,000ms
              </option>
              <option value="5000" key="5000ms">
                5,000ms
              </option>
              <option value="10000" key="10000ms">
                10,000ms
              </option>
              <option value="0" key="0ms">
                0ms
              </option>
            </Select>
          </FormGroup>
        </Layout>

        <Layout flexGrow={0} margin={{ right: 2 }}>
          <FormGroup label="TransitionType [Enter]">
            <Select
              value={this.state.transitionType[0]}
              onChange={this.updateTransitionTypeEnter}
            >
              {Object.keys(TransitionType).map((key: string) => (
                <option
                  value={TransitionType[key as keyof typeof TransitionType]}
                  key={TransitionType[key as keyof typeof TransitionType]}
                >
                  {key}
                </option>
              ))}
            </Select>
          </FormGroup>
        </Layout>

        <Layout flexGrow={0} margin={{ right: 2 }}>
          <FormGroup label="TransitionType [Exit]">
            <Select
              value={this.state.transitionType[1]}
              onChange={this.updateTransitionTypeExit}
            >
              {Object.keys(TransitionType).map((key: string) => (
                <option
                  value={TransitionType[key as keyof typeof TransitionType]}
                  key={TransitionType[key as keyof typeof TransitionType]}
                >
                  {key}
                </option>
              ))}
            </Select>
          </FormGroup>
        </Layout>

        <Layout flexGrow={0} margin={{ right: 2 }}>
          <FormGroup label="Box Size">
            <Range
              min={100}
              max={600}
              value={this.state.boxSize.toString()}
              onChange={this.updateBoxSize}
              disabled
            />
          </FormGroup>
        </Layout>
      </Layout>
    );
  }

  private updateLayout = (e: FormEvent<HTMLSelectElement>) => {
    this.setState({
      layout: e.currentTarget.value as DemoLayoutType,
    });
  };

  private updateDirection = (e: FormEvent<HTMLSelectElement>) => {
    this.setState({
      direction: e.currentTarget.value as FlexDirection,
    });
  };

  private updateDuration = (e: FormEvent<HTMLSelectElement>) => {
    if (e.currentTarget.value.match(/^[0-9]+$/)) {
      this.setState({
        duration: Number(e.currentTarget.value),
      });
    } else {
      this.setState({
        duration: e.currentTarget.value as TransitionDuration,
      });
    }
  };

  private updateTransitionTypeEnter = (e: FormEvent<HTMLSelectElement>) => {
    this.setState({
      transitionType: [
        e.currentTarget.value as TransitionType,
        this.state.transitionType[1] as TransitionType,
      ],
    });
  };

  private updateTransitionTypeExit = (e: FormEvent<HTMLSelectElement>) => {
    this.setState({
      transitionType: [
        this.state.transitionType[0] as TransitionType,
        e.currentTarget.value as TransitionType,
      ],
    });
  };

  private updateBoxSize = (e: FormEvent<HTMLInputElement>) => {
    this.setState({
      boxSize: Number.parseInt(e.currentTarget.value, undefined),
    });
  };
}
