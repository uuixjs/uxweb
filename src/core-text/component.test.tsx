import { Color, FontSize, FontWeight, VerticalAlign } from "../layout";
import { CoreText, CoreTextProps } from "./component";
import { DarkThemeMap, StaticTokenMap } from "lib/ui-tokens";
import {
  LineHeight,
  TextDecoration,
  TextTransform,
  TextType,
  WhiteSpace,
  WordBreak,
} from "./types";
import { mount, shallow } from "enzyme";

import { Enum } from "tachyon-utils-ts";
import { HTMLProps } from "react";
import { setupShallowTest } from "../../tests/helpers";

function getOptionalProps(): CoreTextProps &
  Pick<HTMLProps<HTMLElement>, "id" | "className"> {
  return {
    align: VerticalAlign.TextBottom,
    bold: true,
    breakpointExtraSmall: { fontSize: FontSize.Size1 },
    breakpointSmall: { fontSize: FontSize.Size2 },
    breakpointMedium: { fontSize: FontSize.Size3 },
    breakpointLarge: { fontSize: FontSize.Size4 },
    breakpointExtraLarge: { fontSize: FontSize.Size5 },
    breakpointExtraExtraLarge: { fontSize: FontSize.Size6 },
    className: "test-classname",
    color: Color.Error,
    decoration: TextDecoration.Strikethrough,
    ellipsis: true,
    fontSize: FontSize.Size3,
    fontWeight: FontWeight.SemiBold,
    id: "1234",
    italic: true,
    lineHeight: LineHeight.Body,
    lines: 2,
    title: "Test title",
    transform: TextTransform.Uppercase,
    type: TextType.H3,
    wordBreak: WordBreak.KeepAll,
    whiteSpace: WhiteSpace.NoWrap,
  };
}

const setupShallow = setupShallowTest(CoreText);

describe("CoreText", () => {
  it("renders core text component with required props", () => {
    const { wrapper } = setupShallow();
    expect(wrapper).toMatchSnapshot();
  });

  it("renders core text component with all props as an h3", () => {
    const { wrapper } = setupShallow(getOptionalProps());
    expect(wrapper).toMatchSnapshot();
  });

  it("passes aria props to the element", () => {
    const wrapper = shallow(<CoreText aria-details="id" />);
    const element = wrapper.find("p").first();
    expect(element.prop("aria-details")).toBe("id");
  });

  it("passes data- props to the element", () => {
    const wrapper = mount(<CoreText data-foobar="test">Hello World</CoreText>);
    const domNode = wrapper.find("p").hostNodes();
    expect(domNode).toHaveProp("data-foobar");
    expect(domNode.prop("data-foobar")).toEqual("test");
  });

  it("sets title to children when ellipsis is string", () => {
    const wrapper = shallow(<CoreText ellipsis>Ellipsized Text</CoreText>);
    const element = wrapper.find("p").first();
    expect(element.prop("title")).toBe("Ellipsized Text");
  });

  it("allows custom title when ellipsis is string", () => {
    const wrapper = shallow(
      <CoreText title="custom title" ellipsis>
        Ellipsized Text
      </CoreText>,
    );
    const element = wrapper.find("p").first();
    expect(element.prop("title")).toBe("custom title");
  });

  it("allows title to be unset when ellipsis is defined", () => {
    const wrapper = shallow(
      <CoreText title="" ellipsis>
        Ellipsized Text
      </CoreText>,
    );
    const element = wrapper.find("p").first();
    expect(element.prop("title")).toBe("");
  });

  it("renders a p tag by default", () => {
    const wrapper = shallow(<CoreText>Hello World</CoreText>);
    expect(wrapper.find("p")).toExist();
  });

  it.each(Enum.values(TextType))(
    "renders %s tag via the type prop",
    (tagName) => {
      const wrapper = shallow(<CoreText type={tagName}>Hello World</CoreText>);
      expect(wrapper.find(tagName)).toExist();
    },
  );

  it.each(Enum.values(TextType))(
    "renders %s tag via the as prop",
    (tagName) => {
      const wrapper = shallow(
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        <CoreText as={tagName as any}>Hello World</CoreText>,
      );
      expect(wrapper.find(tagName)).toExist();
    },
  );

  it("sets vertical-align css via the special `align` prop", () => {
    const wrapper = mount(
      <CoreText align={VerticalAlign.Middle}>Hello World</CoreText>,
    );
    const text = wrapper.find("p");
    expect(text).toHaveStyleRule("vertical-align", "middle !important");
  });

  it("sets font-weight css via the special `bold` prop", () => {
    const wrapper = mount(<CoreText bold>Hello World</CoreText>);
    const text = wrapper.find("p");
    expect(text).toHaveStyleRule(
      "font-weight",
      `${StaticTokenMap["font-weight-semibold"]} !important`,
    );
  });

  describe("ellipsis", () => {
    it("sets the right css when ellipsis = true", () => {
      const wrapper = mount(<CoreText lines={1}>Hello World</CoreText>);

      const text = wrapper.find("p");
      expect(text).toHaveStyleRule("overflow: hidden");
      expect(text).toHaveStyleRule("text-overflow: ellipsis");
      expect(text).toHaveStyleRule("white-space", "nowrap");
    });

    it("sets the right css when lines = 1", () => {
      const wrapper = mount(<CoreText lines={1}>Hello World</CoreText>);

      const text = wrapper.find("p");
      expect(text).toHaveStyleRule("overflow: hidden");
      expect(text).toHaveStyleRule("text-overflow: ellipsis");
      expect(text).toHaveStyleRule("white-space", "nowrap");
    });

    it("sets the right css when lines = 2", () => {
      const wrapper = mount(<CoreText lines={2}>Hello World</CoreText>);

      const text = wrapper.find("p");
      expect(text).toHaveStyleRule("overflow: hidden !important");
      expect(text).toHaveStyleRule("text-overflow: ellipsis !important");
      expect(text).toHaveStyleRule("white-space", "normal !important");
      expect(text).toHaveStyleRule("display", "-webkit-box");
      expect(text).toHaveStyleRule("-webkit-line-clamp", "2");
      expect(text).toHaveStyleRule("-webkit-box-orient", "vertical");
    });

    it("sets the right css when lines = 2 and ellipsis = true", () => {
      // There is an issue with jest `toHaveStyleRule` with multiple css props with the same name.
      const wrapper = mount(
        <CoreText ellipsis lines={2}>
          Hello World
        </CoreText>,
      );

      const text = wrapper.find("p");
      expect(text).toHaveStyleRule("overflow: hidden !important");
      expect(text).toHaveStyleRule("text-overflow: ellipsis !important");
      expect(text).toHaveStyleRule("white-space", "normal !important");
      expect(text).toHaveStyleRule("display", "-webkit-box");
      expect(text).toHaveStyleRule("-webkit-line-clamp", "2");
      expect(text).toHaveStyleRule("-webkit-box-orient", "vertical");
    });
  });

  it("sets css for CoreText specific props", () => {
    const wrapper = mount(
      <CoreText
        decoration={TextDecoration.Strikethrough}
        italic
        lineHeight={LineHeight.Heading}
        tabularNums
        transform={TextTransform.Uppercase}
        wordBreak={WordBreak.Normal}
        lines={2}
      >
        Hello World
      </CoreText>,
    );

    const text = wrapper.find("p");

    expect(text).toHaveStyleRule("text-decoration", "line-through !important");
    expect(text).toHaveStyleRule("font-style", "italic !important");
    expect(text).toHaveStyleRule(
      "line-height",
      `${StaticTokenMap["line-height-heading"]} !important`,
    );
    expect(text).toHaveStyleRule("font-variant-numeric", "tabular-nums");
    expect(text).toHaveStyleRule("text-transform", "uppercase !important");
    expect(text).toHaveStyleRule("word-break", "normal !important");
  });

  it("sets css for Layout specific props with breakpoints", () => {
    const wrapper = mount(
      <CoreText
        color={Color.Alt}
        fontSize={FontSize.Size4}
        breakpointLarge={{ fontSize: FontSize.Size5 }}
      >
        Hello World
      </CoreText>,
    );

    const text = wrapper.find("p");

    expect(text).toHaveStyleRule(
      "color",
      expect.stringContaining(DarkThemeMap["color-text-alt"]),
    );
    expect(text).toHaveStyleRule(
      "font-size",
      expect.stringContaining(StaticTokenMap["font-size-4"]),
    );
    expect(text).toHaveStyleRule(
      "font-size",
      expect.stringContaining(StaticTokenMap["font-size-5"]),
      { media: `screen and (min-width:${StaticTokenMap["break-lg"]})` },
    );
  });

  it("does not pass through props to underlying DOM node", () => {
    const wrapper = mount(
      <CoreText
        color={Color.Base}
        align={VerticalAlign.Middle}
        bold
        type={TextType.Span}
      >
        Hello World
      </CoreText>,
    );
    const domNode = wrapper.find("span").hostNodes();
    expect(domNode).not.toHaveProp("color");
    expect(domNode).not.toHaveProp("align");
    expect(domNode).not.toHaveProp("bold");
    expect(domNode).not.toHaveProp("type");
  });

  it("renders children and works with enzyme text()", () => {
    const wrapper = shallow(
      <CoreText type={TextType.Span}>Hello World</CoreText>,
    );

    expect(wrapper.text()).toEqual("Hello World");
  });

  it("typescript disallows Layout props", () => {
    // @ts-expect-error The margin prop is not part of the CoreTextProps interface
    shallow(<CoreText margin={1}>Hello World</CoreText>);
  });
});
