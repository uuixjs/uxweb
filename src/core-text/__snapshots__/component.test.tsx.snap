// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoreText renders core text component with all props as an h3 1`] = `
.c0 {
  color: #ff8280 !important;
  -webkit-text-decoration: line-through !important;
  text-decoration: line-through !important;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
  font-size: 2.4rem !important;
  font-weight: 600 !important;
  font-style: italic !important;
  line-height: 1.5 !important;
  overflow: hidden !important;
  text-overflow: ellipsis !important;
  white-space: normal !important;
  display: -webkit-box;
  -webkit-line-clamp: 2;
  -webkit-box-orient: vertical;
  text-transform: uppercase !important;
  word-break: keep-all !important;
  white-space: nowrap !important;
  vertical-align: text-bottom !important;
}

@media screen and (min-width:480px) {
  .c0 {
    font-size: 5.4rem !important;
  }
}

@media screen and (min-width:768px) {
  .c0 {
    font-size: 3.6rem !important;
  }
}

@media screen and (min-width:1024px) {
  .c0 {
    font-size: 2.4rem !important;
  }
}

@media screen and (min-width:1200px) {
  .c0 {
    font-size: 1.8rem !important;
  }
}

@media screen and (min-width:1440px) {
  .c0 {
    font-size: 1.4rem !important;
  }
}

@media screen and (min-width:1600px) {
  .c0 {
    font-size: 1.3rem !important;
  }
}

<h3
  className="c0 test-classname"
  id="1234"
  title="Test title"
/>
`;

exports[`CoreText renders core text component with required props 1`] = `
<p
  className=""
/>
`;
